File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: list[int], target: int) -> list[int]:
        def findLeft():
            lo, hi = 0, len(nums) - 1
            idx = -1
            while lo <= hi:
                mid = (lo + hi) // 2
                if nums[mid] >= target:
                    hi = mid - 1
                else:
                    lo = mid + 1
                if nums[mid] == target:
                    idx = mid
            return idx
        
        def findRight():
            lo, hi = 0, len(nums) - 1
            idx = -1
            while lo <= hi:
                mid = (lo + hi) // 2
                if nums[mid] <= target:
                    lo = mid + 1
                else:
                    hi = mid - 1
                if nums[mid] == target:
                    idx = mid
            return idx
        
        return [findLeft(), findRight()]

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        return new int[]{findLeft(nums, target), findRight(nums, target)};
    }
    
    private int findLeft(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1, idx = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] >= target) {
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
            if (nums[mid] == target) idx = mid;
        }
        return idx;
    }
    
    private int findRight(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1, idx = -1;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] <= target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
            if (nums[mid] == target) idx = mid;
        }
        return idx;
    }
}
